resources/media-controls.js

```diff
@@ -0,0 +1,381 @@
+      });
+
+      // Init position duration box.
+      const positionTextNode = this.elements.positionText;
+      const durationSpan = this.elements.duration;
+      const durationFormat = durationSpan.textContent;
+      const positionFormat = positionTextNode.textContent;
+
+      durationSpan.classList.add("duration");
+      durationSpan.setAttribute("role", "none");
+
+      Object.defineProperties(this.elements.positionDurationBox, {
+        durationSpan: {
+          value: durationSpan
+        },
+        position: {
+          get: () => {
+            return positionTextNode.textContent;
+          },
+          set: v => {
+            positionTextNode.textContent = positionFormat.replace("#1", v);
+          }
+        },
+        duration: {
+          get: () => {
+            return durationSpan.textContent;
+          },
+          set: v => {
+            durationSpan.textContent = v ? durationFormat.replace("#2", v) : "";
+          }
+        },
+        show: {
+          value: (currentTime, duration) => {
+            const self = this.elements.positionDurationBox;
+            if (self.position != currentTime) {
+              self.position = currentTime;
+            }
+            if (self.duration != duration) {
+              self.duration = duration;
+            }
+            self.classList.remove("hidden");
+          }
+        }
+      });
+
+      // Add event listeners.
+      this.mediaEvents = [
+        "play",
+        "pause",
+        "ended",
+        "volumechange",
+        "loadeddata",
+        "loadstart",
+        "timeupdate",
+        "progress",
+        "playing",
+        "waiting",
+        "canplay",
+        "canplaythrough",
+        "seeking",
+        "seeked",
+        "emptied",
+        "loadedmetadata",
+        "error",
+        "suspend"
+      ];
+      this.mediaEvents.forEach(event => {
+        this.media.addEventListener(event, this);
+      });
+
+      this.controlEvents = [
+        { el: this.elements.playPauseButton, type: "click" },
+        { el: this.elements.volumeSwitch, type: "click" },
+        { el: this.elements.volumeLevel, type: "input" }
+      ];
+      this.controlEvents.forEach(({ el, type }) => {
+        el.addEventListener(type, this);
+      });
+
+      // Create state transitions.
+      //
+      // It exposes one method per transition. i.e. this.pause(), this.play(), etc.
+      // For each transition, we check that the transition is possible and call
+      // the `onStateChange` handler.
+      for (let name in TRANSITIONS) {
+        if (!TRANSITIONS.hasOwnProperty(name)) {
+          continue;
+        }
+        this[name] = () => {
+          const from = this.state;
+
+          // Checks if the transition is valid in the current state.
+          if (!TRANSITIONS[name][from]) {
+            const error = `Transition "${name}" invalid for the current state "${from}"`;
+            console.error(error);
```

Comment:
Let's file a follow-up issue to expose a privileged api for logging an internal error instead.

Reply:
https://github.com/servo/servo/issues/23826